src/java.desktop/share/classes/sun/swing/FilePane.java

```diff
@@ -1189,19 +1190,30 @@ public Component getTableCellRendererComponent(JTable table, Object value,
                 Icon icon = chooser.getIcon(file);
                 setIcon(icon);
 
-            } else if (value instanceof Long) {
-                long len = ((Long) value) / 1024L;
+            } else if (value instanceof Long len) {
                 if (listViewWindowsStyle) {
-                    text = MessageFormat.format(kiloByteString, len + 1);
-                } else if (len < 1024L) {
-                    text = MessageFormat.format(kiloByteString, (len == 0L) ? 1L : len);
+                    if (len == 0) {
+                        text = MessageFormat.format(kiloByteString, len);
+                    } else {
+                        len /= 1000L;
+                        text = MessageFormat.format(kiloByteString, len + 1);
+                    }
+                } else if (len < 1000L) {
+                    text = MessageFormat.format(kiloByteString, (len==0 ? 0L : 1L));
                 } else {
-                    len /= 1024L;
-                    if (len < 1024L) {
-                        text = MessageFormat.format(megaByteString, len);
+                    double kbVal = formatToDoubleValue(len);
+                    len = (long)kbVal;
+                    if (kbVal < baseFileSize) {
+                        text = MessageFormat.format(kiloByteString, kbVal);
                     } else {
-                        len /= 1024L;
-                        text = MessageFormat.format(gigaByteString, len);
+                        double mbVal = formatToDoubleValue(len);
+                        len = (long)mbVal;
+                        if (mbVal < baseFileSize) {
+                            text = MessageFormat.format(megaByteString, mbVal);
+                        } else {
+                            double gbVal = formatToDoubleValue(len);
+                            text = MessageFormat.format(gigaByteString, gbVal);
```

Comment:
should we also handle "{0} TB" cases?

Reply:
Files having "TB" sizes may not be handled now.